--- /home/xiaobai/Downloads/youtube-dl/3/lib/python3.8/site-packages/youtube_dl-2021.2.22-py3.8.egg/youtube_dl/downloader/fragment.py	2021-02-28 05:48:55.468545393 +0800
+++ /home/xiaobai/Downloads/youtube-dl/youtube-dl-patch/downloader/fragment.py	2021-02-28 06:16:58.846847501 +0800
@@ -13,6 +13,8 @@
     sanitized_Request,
 )
 
+# hole patch:                                                                                                                                        
+from ..utils import patch_get_max_path
 
 class HttpQuietDownloader(HttpFD):
     def to_screen(self, *args, **kargs):
@@ -96,7 +98,10 @@
         frag_index_stream.close()
 
     def _download_fragment(self, ctx, frag_url, info_dict, headers=None):
-        fragment_filename = '%s-Frag%d' % (ctx['tmpfilename'], ctx['fragment_index'])
+
+        fragment_filename = patch_get_max_path( ctx['tmpfilename'], '-Frag' + str(ctx['fragment_index']) ) #hole patch
+        #fragment_filename = '%s-Frag%d' % (ctx['tmpfilename'], ctx['fragment_index']) # orig
+
         fragment_info_dict = {
             'url': frag_url,
             'http_headers': headers or info_dict.get('http_headers'),
